src/wayland/buffer/mod.rs

```diff
@@ -0,0 +1,139 @@
+        self.buffer == other.buffer
+    }
+}
+
+impl ManagedBuffer {
+    /// Initializes a buffer, associating some user data with the buffer. This function causes the buffer to
+    /// managed by Smithay.
+    ///
+    /// The data associated with the buffer may obtained using [`BufferManager::buffer_data`].
+    pub fn init_buffer<D, T>(init: &mut DataInit<'_, D>, buffer: New<WlBuffer>, data: T) -> ManagedBuffer
+    where
+        D: BufferHandler + 'static,
+        T: Send + Sync + 'static,
+    {
+        let data = Arc::new(BufferData { data: Box::new(data) });
+        let buffer = init.custom_init(buffer, data.clone());
+
+        ManagedBuffer {
+            buffer: buffer.id(),
+            data,
+        }
+    }
+
+    /// Creates a [`Buffer`] from a [`WlBuffer`].
+    ///
+    /// This function returns [`Err`] if the buffer is not managed by Smithay (such as an EGL buffer).
+    pub fn from_buffer(
+        buffer: &WlBuffer,
+        dh: &mut DisplayHandle<'_>,
+    ) -> Result<ManagedBuffer, UnmanagedResource> {
+        let data = dh
+            .get_object_data(buffer.id())
+            .map_err(|_| UnmanagedResource)?
+            .downcast::<BufferData>()
+            .map_err(|_| UnmanagedResource)?;
+
+        Ok(ManagedBuffer {
+            buffer: buffer.id(),
+            data,
+        })
+    }
+
+    /// Returns a reference to the underlying [`WlBuffer`].
+    pub fn buffer(&self, dh: &mut DisplayHandle<'_>) -> Result<WlBuffer, InvalidId> {
+        WlBuffer::from_id(dh, self.buffer.clone())
+    }
+
+    /// Sends a `release` event, indicating the buffer is no longer in use by the compositor, meaning the
+    /// client is free to reuse or destroy the buffer.
+    pub fn release(&self, dh: &mut DisplayHandle<'_>) -> Result<(), InvalidId> {
+        self.buffer(dh)?.release(dh);
+        Ok(())
+    }
+
+    /// Returns the data associated with the buffer.
+    ///
+    /// This function is intended for abstractions to obtain the buffer data they store. Users should use the
+    /// functions provided by the buffer abstractions to obtain the data. For example, data associated with an
+    /// shm buffer should be obtained using [`with_buffer_contents`](crate::wayland::shm::with_buffer_contents)
+    /// instead of specifying the type of the data in the `T` generic for this function.
+    pub fn buffer_data<T>(&self) -> Option<&T>
+    where
+        T: Send + Sync + 'static,
+    {
+        <dyn Any>::downcast_ref::<T>(&*self.data.data)
+    }
+}
+
+#[derive(Debug)]
+struct BufferData {
+    data: Box<(dyn Any + Send + Sync + 'static)>,
+}
+
+impl<D> ObjectData<D> for BufferData
+where
+    D: BufferHandler + 'static,
+{
+    fn request(
+        self: Arc<Self>,
+        _: &mut Handle<D>,
+        data: &mut D,
+        _: ClientId,
+        msg: Message<ObjectId>,
+    ) -> Option<Arc<dyn ObjectData<D>>> {
+        // WlBuffer has a single request which is a destructor.
+        assert_eq!(msg.opcode, 0);
```

Comment:
Let's make that a `debug_assert!()`, no need to pollute release code with such checks.